packages/internals/src/views/handleViewsIO.ts

```diff
@@ -0,0 +1,117 @@
+import * as T from 'fp-ts/lib/Task'
+import * as TE from 'fp-ts/lib/TaskEither'
+import path from 'path'
+import { match } from 'ts-pattern'
+
+import {
+  createDirIfNotExists,
+  getFilesInDir,
+  getFoldersInDir,
+  removeDir,
+  removeFile,
+  writeFile,
+} from '../utils/fs-functional'
+
+export interface IntrospectionViewDefinition {
+  // The database or schema where the view is located
+  schema: string
+
+  // The name of the view
+  name: string
+
+  // The database SQL query that defines the view
+  definition: string
+}
+
+type HandleViewsIOParams = {
+  views: IntrospectionViewDefinition[]
+  schemaPath: string
+}
+
+/**
+ * For any given view definitions, the CLI must either create or update the corresponding view definition files
+ * in the file system, in `${path.dirname(schemaPath)}/views/{viewDbSchema}/{viewName}.sql`.
+ * If some other files or folders exist within the `views` directory, the CLI must remove them.
+ * These files and folders are deleted silently.
+ */
+export async function handleViewsIO({ views, schemaPath }: HandleViewsIOParams): Promise<void> {
+  const prismaDir = path.dirname(schemaPath)
+  const viewsDir = path.join(prismaDir, 'views')
+
+  // collect the newest view definitions
+  const viewEntries = views.map(({ schema, ...rest }) => {
+    const viewDir = path.join(viewsDir, schema)
+    return [viewDir, rest] as const
+  })
+
+  // collect the paths to the view directories (identified by their db schema name)
+  const viewPathsToWrite = viewEntries.map(([viewDir, _]) => viewDir)
+
+  // collect the files paths and content for the views' SQL definitions
+  const viewsFilesToWrite = viewEntries.map(([viewDir, { name, definition }]) => {
+    const viewFile = path.join(viewDir, `${name}.sql`)
+    return { path: viewFile, content: definition } as const
+  })
+
+  const updateDefinitionsInViewsDirPipeline = pipe(
+    // create the views directory, idempotently
+    createDirIfNotExists(viewsDir),
+
+    // create the view directories, idempotently and concurrently, collapsing the possible errors
+    TE.chainW(() => TE.traverseArray(createDirIfNotExists)(viewPathsToWrite)),
+
+    // write the view definitions in the directories just created, idempotently and concurrently, collapsing the possible errors
+    TE.chainW(() => TE.traverseArray(writeFile)(viewsFilesToWrite)),
+
+    // remove any view directories related to schemas that no longer exist, concurrently, collapsing the possible errors
+    TE.chainW(() =>
+      pipe(
+        getFoldersInDir(viewsDir),
+        T.chain((directoriesInViewsDir) => {
+          const viewDirsToRemove = directoriesInViewsDir.filter((dir) => !viewPathsToWrite.includes(dir))
+          return TE.traverseArray(removeDir)(viewDirsToRemove)
+        }),
+      ),
+    ),
+
+    // remove any other files in the views directory beyond the ones just created, concurrently, collapsing the possible errors
+    TE.chainW(() =>
+      pipe(
+        getFilesInDir(viewsDir),
```

Comment:
Should it maybe only remove stray `.sql` files? Doesn't sound right to remove `README.md` that someone put it or something like that.